src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -570,82 +603,265 @@ namespace nbl::ext::imgui
+						scissor.offset.y = static_cast<int32_t>(clipRectangle.y);
+						scissor.extent.width = static_cast<uint32_t>(clipRectangle.z - clipRectangle.x);
+						scissor.extent.height = static_cast<uint32_t>(clipRectangle.w - clipRectangle.y);
+						
+						return scissor;
+					}
+				}
 
-			if (static_cast<bool>(vertexBuffer) == false || vertexBuffer->getSize() < vertexSize)
+				bool prepass(const ImDrawCmd* cmd) const
+				{
+					const auto rectangle = getClipRectangle(cmd);
+
+					return prepass(rectangle);
+				}
+
+				bool prepass(const ImVec4& clipRectangle) const
+				{
+					return clipRectangle.x < framebuffer.x && clipRectangle.y < framebuffer.y && clipRectangle.z >= 0.0f && clipRectangle.w >= 0.0f;
+				}
+
+			} clip { .off = drawData->DisplayPos, .scale = drawData->FramebufferScale, .framebuffer = { frameBufferWidth, frameBufferHeight } };
+			
+			struct TRS
 			{
-				vertexBuffer = m_device->createBuffer(std::move(vertexCreationParams));
+				core::vector2df_SIMD scale;
+				core::vector2df_SIMD translate;
 
-				video::IDeviceMemoryBacked::SDeviceMemoryRequirements memReq = vertexBuffer->getMemoryReqs();
-				memReq.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
-				auto memOffset = m_device->allocate(memReq, vertexBuffer.get());
-				assert(memOffset.isValid());
-			}
+				core::vector2df_SIMD toNDC(core::vector2df_SIMD in) const
+				{
+					return in * scale + translate;
+				}
+			};
 
-			IGPUBuffer::SCreationParams indexCreationParams = {};
-			indexCreationParams.usage = nbl::core::bitflag(nbl::asset::IBuffer::EUF_VERTEX_BUFFER_BIT) | nbl::asset::IBuffer::EUF_INDEX_BUFFER_BIT
-				| nbl::asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
-			indexCreationParams.size = indexSize;
+			const TRS trs = [&]() 
+			{
+				TRS retV;
 
-			auto & indexBuffer = m_indexBuffers[frameIndex];
+				retV.scale = core::vector2df_SIMD{ 2.0f / drawData->DisplaySize.x , 2.0f / drawData->DisplaySize.y };
+				retV.translate = core::vector2df_SIMD { -1.0f, -1.0f } - core::vector2df_SIMD{ drawData->DisplayPos.x, drawData->DisplayPos.y } * trs.scale;
 
-			if (static_cast<bool>(indexBuffer) == false || indexBuffer->getSize() < indexSize)
+				return std::move(retV);
+			}();
+			
+			struct
 			{
-				indexBuffer = m_device->createBuffer(std::move(indexCreationParams));
+				size_t vBufferSize = {}, iBufferSize = {},	// sizes in bytes
+				vPadding = {};								// indicies can break our alignment so small padding may be required before vertices in memory, vertices' offset must be multiple of 4
 
-				video::IDeviceMemoryBacked::SDeviceMemoryRequirements memReq = indexBuffer->getMemoryReqs();
-				memReq.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
-				auto memOffset = m_device->allocate(memReq, indexBuffer.get());
+				std::vector<VkDrawIndexedIndirectCommand> indirects;
+				std::vector<PerObjectData> elements;
```

Comment:
when you use the `StreamingTransientBufferST` you'll be able to get pointers you can write directly to without having to build a temporary `std::vector` first

Reply:
resolved